shapes/main.cpp

```diff
@@ -1,80 +1,112 @@
-#include <string>
-#include <memory>
-#include "Shape.hpp"
+#include "Circle.hpp"
 #include "Rectangle.hpp"
+#include "Shape.hpp"
 #include "Square.hpp"
-#include "Circle.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <vector>
 
 using namespace std;
 
 using Collection = vector<shared_ptr<Shape>>;
 
-bool sortByArea(shared_ptr<Shape> first, shared_ptr<Shape> second)
-{
-    if(first == nullptr || second == nullptr)
+auto sortByArea = [](shared_ptr<Shape> first, shared_ptr<Shape> second) {
+    if (first == nullptr || second == nullptr)
         return false;
     return (first->getArea() < second->getArea());
-}
+};
 
-bool perimeterBiggerThan20(shared_ptr<Shape> s)
-{
-    if(s)
+auto perimeterBiggerThan20 = [](shared_ptr<Shape> s) {
+    if (s)
         return (s->getPerimeter() > 20);
     return false;
-}
+};
 
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
+auto areaLessThanX = [x{10}](shared_ptr<Shape> s) {
+    if (s)
+        return (s->getArea() < x);
     return false;
-}
+};
 
 void printCollection(const Collection& collection)
 {
-    for (const auto & it : collection)
+    for (const auto& it : collection)
         if (it)
             it->print();
 }
 
 void printAreas(const Collection& collection)
 {
-    for (const auto & it : collection)
+    for (const auto& it : collection)
         if (it)
             cout << it->getArea() << std::endl;
 }
 
 void findFirstShapeMatchingPredicate(const Collection& collection,
-                                     bool (*predicate)(shared_ptr<Shape> s),
+                                     std::function<bool(shared_ptr<Shape>)> predicate,
                                      std::string info)
 {
     auto iter = std::find_if(collection.begin(), collection.end(), predicate);
-    if(*iter != nullptr)
-    {
+    if (*iter != nullptr) {
         cout << "First shape matching predicate: " << info << endl;
         (*iter)->print();
-    }
-    else
-    {
+    } else {
         cout << "There is no shape matching predicate " << info << endl;
     }
 }
 
-constexpr int fibo(int n) {
-    if (n<=2) {
+constexpr int fibo(int n)
+{
+    if (n <= 2) {
         return 1;
     } else {
         return fibo(n - 1) + fibo(n - 2);
     }
 }
 
+template <typename ShapeSubClass, std::enable_if_t<std::is_base_of_v<Shape, ShapeSubClass>, int> = 0>
+void insertOnlyShapeSubclasses(Collection& collection, const ShapeSubClass& subclass)
+{
+    collection.emplace_back(std::make_shared<ShapeSubClass>(subclass));
+}
+
+std::map<std::shared_ptr<Shape>, double> shapePerimeter{
+    {std::make_shared<Circle>(3.0), std::make_shared<Circle>(3.0)->getPerimeter()},
+    {std::make_shared<Rectangle>(3.0, 5.0), std::make_shared<Rectangle>(3.0, 5.0)->getPerimeter()},
+    {std::make_shared<Square>(4.0), std::make_shared<Square>(4.0)->getPerimeter()},
```

Comment:
that's not a good thing. You have created 6 shapes here. 3 of them are temporary and you copy only values of perimeter from them. There are no memory leaks of course, but dynamic memory allocation and deallocation is not free. You should move this map into main and create shared_pts before you insert them into map.